ruby/src/main/java/com/google/protobuf/jruby/RubyMap.java

```diff
@@ -189,8 +190,25 @@ public IRubyObject index(ThreadContext context, IRubyObject key) {
      */
     @JRubyMethod(name = "==")
     public IRubyObject eq(ThreadContext context, IRubyObject _other) {
-        if (_other instanceof RubyHash)
-            return toHash(context).op_equal(context, _other);
+        if (_other instanceof RubyHash){
+            // quick check to see if they are equivalent
+            IRubyObject matched = toHash(context).op_equal(context, _other);
+            if (matched == context.runtime.getTrue()) {
+                return context.runtime.getTrue();
+            }
+
+            RubyHash other = (RubyHash)_other;
```

Comment:
the reason we have another step here is that `#toHash` will convert all complex objects, such as `Map`, `Message`, and `RepeatedField` into hashes and arrays.  I believe this is what one would expect when calling `#to_h`. However, that means that if the `==` is comparing against something like what is in `ruby/tests/basic.rb#ln616` will fail unless we do a deeper check.

let me know what you think and if this is too complex and/or the assumptions on `#to_h` are wrong